src/include/btree.i

```diff
@@ -1352,8 +1352,9 @@ __wt_page_can_evict(WT_SESSION_IMPL *session, WT_REF *ref, bool *inmem_splitp)
 
 	/* A truncated page can't be evicted until the truncate completes. */
 	if (ref->page_del != NULL && ref->page_del->txnid != WT_TXN_ABORTED &&
-	    !__wt_txn_visible_all(session,
-	    ref->page_del->txnid, WT_TIMESTAMP_NULL(&ref->page_del->timestamp)))
+	    !(ref->page_del->state == WT_UPDATE_STATE_READY &&
+	    __wt_txn_visible_all(session, ref->page_del->txnid,
+	    WT_TIMESTAMP_NULL(&ref->page_del->timestamp))))
```

Comment:
I've been staring at this, and I think this test is unsafe.

Essentially, we're using `ref->page_del != NULL && ref->page_del->txnid != WT_TXN_ABORTED` as a proxy for `WT_REF.state == WT_REF_DELETED`, which I don't completely understand, but more importantly, I don't think this test is safe because we don't have the page locked down and can race with another thread freeing`ref.page_del` as part of evicting the page.

I think we have to move this test into `__evict_review()`, done after we lock the page down.

If I'm correct, then this code in `__wt_delete_page_rollback()` is probably in play as well.
```
done:   /*
         * Now mark the truncate aborted: this must come last because after
         * this point there is nothing preventing the page from being evicted.
         */
        ref->page_del->state = WT_UPDATE_STATE_READY;
        WT_PUBLISH(ref->page_del->txnid, WT_TXN_ABORTED);
        return (0);
```
My guess is setting `page_del->txnid = WT_TXN_ABORTED` would no longer be useful, but I'd want to stare at it for a few minutes before making that change.

@agorrod, thoughts? (This can obviously be a different ticket.)

Reply:
created a new ticket [WT-3992](https://jira.mongodb.org/browse/WT-3992)